wasm/integration-test/crypto/crypto.c

```diff
@@ -0,0 +1,114 @@
+{
+}
+
+// Exported Functions from `hermes:cardano/event-on-txn`
+void exports_hermes_cardano_event_on_txn_on_cardano_txn(exports_hermes_cardano_event_on_txn_cardano_blockchain_id_t blockchain, uint64_t slot, uint32_t txn_index, exports_hermes_cardano_event_on_txn_cardano_txn_t *txn)
+{
+}
+
+// Exported Functions from `hermes:cardano/event-on-rollback`
+void exports_hermes_cardano_event_on_rollback_on_cardano_rollback(exports_hermes_cardano_event_on_rollback_cardano_blockchain_id_t blockchain, uint64_t slot)
+{
+}
+
+// Exported Functions from `hermes:cron/event`
+bool exports_hermes_cron_event_on_cron(exports_hermes_cron_event_cron_tagged_t *event, bool last)
+{
+  return false;
+}
+
+// Exported Functions from `hermes:init/event`
+bool exports_hermes_init_event_init(void)
+{
+  return false;
+}
+
+// Exported Functions from `hermes:kv-store/event`
+void exports_hermes_kv_store_event_kv_update(hermes_string_t *key, exports_hermes_kv_store_event_kv_values_t *value)
+{
+}
+
+#define HERMES_STRING(x) \
+  (hermes_string_t) { (uint8_t *)x, strlen(x) }
+
+hermes_crypto_api_borrow_bip32_ed25519_t get_or_add_resource()
+{
+  hermes_string_t mnemonic_string = HERMES_STRING("prevent company field green slot measure chief hero apple task eagle sunset endorse dress seed");
+  hermes_crypto_api_mnemonic_phrase_t mnemonic = {.ptr = &mnemonic_string, .len = 1};
+
+  hermes_string_t passphrase_string = {.ptr = NULL, .len = 0};
+  hermes_crypto_api_passphrase_t passphrase = {.ptr = &passphrase_string, .len = 0};
+
+  hermes_crypto_api_own_bip32_ed25519_t resource = hermes_crypto_api_constructor_bip32_ed25519(&mnemonic, NULL);
+  hermes_crypto_api_borrow_bip32_ed25519_t borrow_resource = hermes_crypto_api_borrow_bip32_ed25519(resource);
+  return borrow_resource;
+}
+
+bool generate_mnemonic()
+{
+  hermes_string_t prefix_data = HERMES_STRING("project");
+
+  hermes_crypto_api_prefix_t prefix = {.ptr = &prefix_data, .len = 1};
+  hermes_string_t language = HERMES_STRING("English");
+
+  hermes_crypto_api_mnemonic_phrase_t ret;
+  hermes_crypto_api_errno_t err;
+
+  hermes_crypto_api_generate_mnemonic(24, &prefix, &language, &ret, &err);
+
+  char *expected_prefix = "project";
+  size_t n = strlen(expected_prefix);
+
+  return ret.ptr != NULL && ret.ptr->ptr != NULL && ret.ptr->len >= n &&
+         strncmp((const char *)ret.ptr->ptr, expected_prefix, n) == 0;
+}
+
+bool get_pubkey()
+{
+  hermes_crypto_api_borrow_bip32_ed25519_t borrow_resource = get_or_add_resource();
+  hermes_crypto_api_bip32_ed25519_public_key_t ret;
+  hermes_crypto_api_method_bip32_ed25519_public_key(borrow_resource, &ret);
+
+  return (ret.f0 == 3986768884739312704ULL) &&
+         (ret.f1 == 9782938079688165927ULL) &&
+         (ret.f2 == 7977656244723921923ULL) &&
+         (ret.f3 == 12587033252467133758ULL);
+}
```

Comment:
```suggestion
bool bip32_ed25519()
{
  // Test public key.
    hermes_crypto_api_borrow_bip32_ed25519_t borrow_resource = get_or_add_resource();
    hermes_crypto_api_bip32_ed25519_public_key_t gen_pub_key_ret;
    hermes_crypto_api_method_bip32_ed25519_public_key(borrow_resource, &gen_pub_key_ret);

    bool public_key_success = (gen_pub_key_ret.f0 == 3986768884739312704) &&
                              (gen_pub_key_ret.f1 == 9782938079688165927ULL) &&
                              (gen_pub_key_ret.f2 == 7977656244723921923) &&
                              (gen_pub_key_ret.f3 == 12587033252467133758ULL);
    if (!public_key_success)
    {
      return false;
    }

    // Test sign and check sig.
    hermes_string_t data_string = {.ptr = (uint8_t *)"test", .len = 4};
    hermes_crypto_api_bstr_t sign_data = {.ptr = &data_string, .len = 1};
    hermes_crypto_api_bip32_ed25519_signature_t gen_sig_ret;

    hermes_crypto_api_method_bip32_ed25519_sign_data(borrow_resource, &sign_data, &gen_sig_ret);
    bool sign_success = hermes_crypto_api_method_bip32_ed25519_check_sig(borrow_resource, &sign_data, &gen_sig_ret);
    if (!sign_success)
    {
      return false;
    }

    // Test derive key.
    hermes_string_t path_string = {.ptr = (uint8_t *)"m/1852'/1815'/0'/2/0", .len = strlen("m/1852'/1815'/0'/2/0")};
    hermes_crypto_api_own_bip32_ed25519_t new_resource = hermes_crypto_api_method_bip32_ed25519_derive(borrow_resource, &path_string);
    return new_resource.__handle == 2;
}
```

Reply:
I deliberately simplified the test.
Multiple tests one after the other here make it difficult to find out what wrong when something breaks, it can be any number of cascading permutations.
We don;t need exhaustive behaviour checks in the wasm, but if you think we need to test these functions, then they need to be individual checks.